### PR TITLE
Blog Post for Autobuild docs for Pull Requests

### PR DESCRIPTION
**Pull request by saadmk11**

### CODE REVIEW

File: pr-builder.rst [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+IRC channel or `readthedocs gitter`_ channel. I am `saadmk11` at IRC and `@saadmk11` at gitter.
+
+Conclusion
+==========
+
+To conclude, I would like to say this was a much needed feature for Read the Docs and also its users.
+This feature will improve our platform and make it a true Continuous Documentation platform.
+I think that many users will benefit from this feature.
+
+We will extend on this feature and integrate it to other platforms like ``GitLab/BitBucket`` in the near future.
```
Comment (davidfischer) [inline pr-builder.rst]:
```suggestion
We will extend on this feature and integrate it to other platforms like GitLab and BitBucket in the near future.
```

File: pr-builder.rst [hunk 2]
```diff
@@ -0,0 +1,93 @@
+.. post:: July 14, 2019
+   :tags: gsoc, pr-builder, continuous-documentation, feature
+   :author: Saadmk11
+   :location: DHA
+
+Autobuild docs for Pull Requests
```
Comment (davidfischer) [inline pr-builder.rst]: We typically use titlecase for blogs and I'd spell out documentation or use something like "Building the Docs".

"Autobuild docs for Pull Requests" -> "Building the Docs in Pull Requests" or "Autobuild Documentation in Pull Requests".

File: pr-builder.rst [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+As a student of Google Summer of Code (GSoC) 2019, I (`Maksudul Haque`_) was tasked with building this feature.
+The Main goal of my project was to make it possible to build documentation when ever a pull/merge request was created,
+and send build status notification to the Providers Status API.
+
+Google Summer of Code
+=====================
+
+Google Summer of Code is a global program where students work with an open source organization
+on a 3 month programming project. **Autobuild docs for Pull Requests** was one of the project ideas on
```
Comment (davidfischer) [inline pr-builder.rst]: If you do change the title of the blog, I'd change this line as well.

File: pr-builder.rst [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+Read the Docs `Project Ideas`_ for Google Summer of Code 2019. I applied for the project and got accepted.
+
+I have started working on this project from May 2019.
+All of my work on this project can be seen in the `Pull Request Builder Board`_.
+
+Background
+==========
+
+Many of our users wanted a way to visualize the documentation update that was made in a pull/merge request.
+They also wanted to know the status of a build ``(passed/failed)`` before merging the pull/merge request to master.
```
Comment (davidfischer) [inline pr-builder.rst]:
```suggestion
They also want to know whether the documentation build will pass before merging the pull/merge request to master.
```

File: pr-builder.rst [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+This would allow users to have more confidence on the pull request
+and make the pull/merge request less likely to break the documentation after merging.
+So, In reply to this Read the Docs `core team`_ selected **Autobuild docs for Pull Requests**
```
Comment (davidfischer) [inline pr-builder.rst]: This sentence is a bit awkward.

File: pr-builder.rst [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+as one of the projects of Google Summer of Code 2019.
+
+Pull Request Builder Features
+=============================
+
+Currently I have implemented these features and working on more upcoming features.
+Some of the major features are as following:
+
+- **Creating External Versions:** We create an ``External Version`` when we receive ``pull_request`` ``opened/reopened``
```
Comment (davidfischer) [inline pr-builder.rst]: These should probably not be monospaced. They aren't code.

Screen Shot 2019-07-15 at 11 17 42 AM

File: pr-builder.rst [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+Conclusion
+==========
+
+To conclude, I would like to say this was a much needed feature for Read the Docs and also its users.
+This feature will improve our platform and make it a true Continuous Documentation platform.
+I think that many users will benefit from this feature.
+
+We will extend on this feature and integrate it to other platforms like ``GitLab/BitBucket`` in the near future.
+We would like to add more features on this. So, It would be Great to have Feedback form the users.
```
Comment (davidfischer) [inline pr-builder.rst]:
```suggestion
If there are additional features you'd like to see for building documentation in pull requests, please `let us know <mailto:team@readthedocs.org>`_.
```

File: pr-builder.rst (hunk 2 above)
Comment (saadmk11) [reply pr-builder.rst]: Autobuild Documentation in Pull Requests sounds nice. Thanks :)

File: pr-builder.rst (hunk 6 above)
Comment (saadmk11) [reply pr-builder.rst]: okay

**Comment by saadmk11** [comment]
> One of the biggest things to highlight though is how users can start using it! Is it just on for their project already? Do they need to do anything? Let's be really clear there. That's how you'll get feedback!

We will most probably deploy this behind a feature flag initially. So, Should we tell the users to ask us to enable this?
Otherwise they don't need anything to Get Started as only opening a pull request will build their docs there is no initial setup needed.

**Comment by davidfischer** [comment]
If it's behind a feature flag, then I say we should do something like "email us to take part in the beta" or something like that.

**Comment by saadmk11** [comment]
@davidfischer Updated the PR. let me know if we can add anything else like screenshots or something :)

**Comment by davidfischer** [comment]
I would definitely add a screenshot! Pictures can really tell the story.

File: pr-builder.rst [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+as one of the projects of Google Summer of Code 2019.
+
+Pull Request Builder Features
+=============================
+
+Currently I have implemented these features and working on more upcoming features.
+Some of the major features are as following:
+
+- **Creating External Versions:** We create an External Version when we receive pull_request opened/reopened
+  webhook event for a project from gitHub and trigger a build for that version.
```
Comment (davidfischer) [inline pr-builder.rst]: Maybe give a small bit of detail here like:

    External versions are short-lived versions for a Read the Docs project that are separate from the project's main documentation.

File: pr-builder.rst [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+So, to achieve this Read the Docs `core team`_ selected **Autobuild docs for Pull Requests**
+as one of the projects of Google Summer of Code 2019.
+
+Pull Request Builder Features
+=============================
+
+Currently I have implemented these features and working on more upcoming features.
+Some of the major features are as following:
+
+- **Creating External Versions:** We create an External Version when we receive pull_request opened/reopened
```
Comment (davidfischer) [inline pr-builder.rst]:
```suggestion
- **Creating External Versions:** We create an external version when we receive a pull request
```

File: pr-builder.rst (hunk 8 above)
Comment (davidfischer) [inline pr-builder.rst]:
```suggestion
  webhook event for a project from GitHub and trigger a build for that version.
```

File: pr-builder.rst [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+  webhook event for a project from gitHub and trigger a build for that version.
+
+- **Synchronizing External Versions:** Whenever there is a new commit on the pull/merge request,
+  we synchronize the external version for that pull/merge request and trigger a new build with the latest changes.
```
Comment (davidfischer) [inline pr-builder.rst]:
```suggestion
  we synchronize the external version for that pull request and trigger a new documentation build with the latest changes.
```

File: pr-builder.rst [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+- **Deleting External Versions:** Whenever the pull/merge request is closed or merged,
+  we delete the external Version associated with that pull/merge request.
+
+- **Warning Banner for pull/merge request documentation:** While building documentation for pull/merge requests
+  we add a warning banner at the top of those documentations to let the users know that
+  this documentation was generated for pull/merge requests and is not the main documentation for the project.
+
+- **Sending Status Notification to Status API:** We send Build status reports to Status API of the provider (e.g. GitHub) of
```
Comment (davidfischer) [inline pr-builder.rst]:
```suggestion
We send build status reports to status API of the provider (e.g. GitHub) for
```

Comment (davidfischer) [reply pr-builder.rst]: Maybe even remove "for the project" from the end of the sentence.

File: pr-builder.rst [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+  the project. When a build is triggered for a pull/merge request we send `state: pending` notification with build URL
+  and after the build has finished we send `state: success` notification if the build succeeded without any error
+  or `state: failure` notification if the build failed with build URL.
```
Comment (davidfischer) [inline pr-builder.rst]: This sentence may be a bit too much detail for a blog post.

File: pr-builder.rst [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+  By going to the build URL provided in the status report users can view the build steps
+  and also see the documentation generated by that build.
+
+**Currently, we only support GitHub and planning to extend to GitLab and BitBucket**
+
+Getting Started
+===============
+
+**Autobuild Documentation in Pull Requests** is currently on Beta testing and only supports Github repositories.
```
Comment (davidfischer) [inline pr-builder.rst]:
```suggestion
Build documentation in pull requests is currently on Beta testing and only supports Github repositories.
```

The formal name of the project is probably not *that* relevant to readers of the blog.

File: pr-builder.rst [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+If you want to dive in and enable this feature for your project you can `email us <mailto:support@readthedocs.org>`_.
```
Comment (davidfischer) [inline pr-builder.rst]: :+1:

**Comment by safwanrahman** [comment]
The blog post looks awesome. Thanks a lot @saadmk11 
Can you add some future improvements of this feature if possible?

**Comment by saadmk11** [comment]
Sure. I'll try my best to add more improvements in the future. :)

**Comment by safwanrahman** [comment]
@saadmk11 I meant Can you include the future improvement scope in the blog post?

File: building-docs-in-pull-requests.rst [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+or failure notification if the build failed.
+By going to the build URL provided in the status report users can view the build steps
+and also see the documentation generated by that build.
+
+**Currently, we only support GitHub and planning to extend to GitLab and BitBucket**
+
+Getting Started
+===============
+
+Build documentation in pull requests is currently on Beta testing and only supports Github repositories.
```
Comment (davidfischer) [inline building-docs-in-pull-requests.rst]:
```suggestion
Building documentation for pull requests is currently in Beta testing and only supports Github repositories.
```

**Comment by saadmk11** [comment]
We should also reference an issue where users can comment to turn this feature on for their projects in the getting started section before we merge it.

File: building-docs-in-pull-requests.rst [hunk 16]
```diff
@@ -0,0 +1,116 @@
+.. post:: July 14, 2019
```
Comment (ericholscher) [inline building-docs-in-pull-requests.rst]: This need to be updated. Let's plan for tomorrow?

```suggestion
.. post:: Aug 14, 2019
```

File: building-docs-in-pull-requests.rst [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+   :tags: gsoc, pr-builder, continuous-documentation, feature
+   :author: Saadmk11
+   :location: DHA
+
+Autobuild Documentation in Pull Requests
```
Comment (ericholscher) [inline building-docs-in-pull-requests.rst]:
```suggestion
GSOC 2019: Autobuild Documentation in Pull Requests
```

I think this is good to mention 👍

File: building-docs-in-pull-requests.rst [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+========================================
```
Comment (ericholscher) [inline building-docs-in-pull-requests.rst]: (Need to update this too, to match :D

File: building-docs-in-pull-requests.rst [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+By going to the build URL provided in the status report users can view the build steps
+and also see the documentation generated by that build.
+
+**Currently, we only support GitHub and planning to extend to GitLab and BitBucket**
+
+Getting Started
+===============
+
+Building documentation for pull requests is currently in Beta testing and only supports Github repositories.
+If you want to dive in and enable this feature for your project you can `email us <mailto:support@readthedocs.org>`_.
```
Comment (ericholscher) [inline building-docs-in-pull-requests.rst]: We should mention that people should include a project URL that they want to enable.

File: building-docs-in-pull-requests.rst [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+
+Future Improvements
+===================
+
+We are planing to extend this feature to other platforms such as GitLab and BitBucket.
+We are also planning to make this feature more customizable through our configaration file (``.readthedocs.yaml``).
+If you have any improvements or features in mind for building documentation in pull requests we would love to know about it.
+Please feel free to `let us know <mailto:team@readthedocs.org>`_.
```
Comment (ericholscher) [inline building-docs-in-pull-requests.rst]: I'd say to file an issue on GitHub here.

**Comment by saadmk11** [comment]
@ericholscher Updated the PR :)

**Comment by ericholscher** [comment]
Whole lot of new editor files that need to be removed :)

**Comment by saadmk11** [comment]
@ericholscher Done! Sorry! Accidentally pushed that :(

**Comment by ericholscher** [comment]
Will plan to merge this tomorrow.

**Comment by saadmk11** [comment]
@ericholscher Added some minor changes.

**Comment by ericholscher** [comment]
👍

**Pull request merged**